### PR TITLE
Added functional and integration tests

### PR DESCRIPTION
**Pull request by adetolacog**
**Before creating a pull request make sure that:**

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)

Please remove this line and everything above and fill the following sections:

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: src/integrationTest/resources/application.yaml [hunk 1]
```diff
@@ -163,8 +163,22 @@ idam:
     url: https://idam-api.aat.platform.hmcts.net
   client:
     id: prl-cos-api
+    baseUrl: https://idam-api.aat.platform.hmcts.net
     redirect_uri: https://manage-case.aat.platform.hmcts.net/oauth2/callback
     secret: ${IDAM_CLIENT_SECRET:xxxxxxx}
+    clientId: xuiwebapp
```
Comment (parvezhmcts) [inline src/integrationTest/resources/application.yaml:169]: are these changes required?

File: src/integrationTest/resources/application.yaml [hunk 2]
```diff
@@ -271,6 +285,12 @@ document:
       fl401_draft_welsh_filename: Draft_DA_Welsh_application
       fl401_c8_welsh_template: FL401-c8-welsh.docx
       fl401_c8_welsh_filename: C8_Document_Welsh.pdf
+      fl401_fl406_draft_template: PRL-GOR-ENG-FL406-Power-Of-Arrest-Draft.docx
```
Comment (parvezhmcts) [inline src/integrationTest/resources/application.yaml:288]: are these changes are part of your story?

File: src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/CafCassControllerFunctionalTest.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        SearchResult expectedCafcassResponse = objectMapper.readValue(cafcassResponseStr, SearchResult.class);
+
+        Mockito.when(postcodeLookupService.isValidNationalPostCode(anyString(), anyString())).thenReturn(true);
+        Mockito.when(coreCaseDataApi.searchCases(anyString(), anyString(), anyString(), anyString())).thenReturn(expectedCafcassResponse);
+
+        MvcResult mvcResult = mockMvc.perform(get(SEARCH_CASE_URL)
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .header("authorisation", AUTH_TEST_TOKEN)
+                        .header("serviceAuthorisation", SERVICE_AUTH_TEST_TOKEN)
+                        .queryParam("start_date", "2022-08-22T10:44:43.49")
```
Comment (parvezhmcts) [inline src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/CafCassControllerFunctionalTest.java:78]: create constants wherever applicable?

**Review by parvezhmcts — COMMENTED** [review]
please check review comments

File: src/integrationTest/resources/application.yaml (hunk 1 above)
Comment (adetolacog) [reply src/integrationTest/resources/application.yaml:169]: Not particularly but added in the process of resolving innitial build issues to mimmick other test application.yaml files

File: src/integrationTest/resources/application.yaml (hunk 2 above)
Comment (adetolacog) [reply src/integrationTest/resources/application.yaml:288]: Added to ensure documents are identical to those within other test areas in the process of resolving build issues.

File: src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/CafCassControllerFunctionalTest.java (hunk 3 above)
Comment (adetolacog) [reply src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/CafCassControllerFunctionalTest.java:78]: Constants already created but now moved into utils/TestConstants Class

File: src/contractTest/java/uk/gov/hmcts/reform/prl/clients/cafcass/CafcassSearchCaseApiConsumerApplication.java [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package uk.gov.hmcts.reform.prl.clients.cafcass;
+
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.cloud.openfeign.EnableFeignClients;
+
+@SpringBootApplication
+@EnableFeignClients(basePackages = {
+        "uk.gov.hmcts.reform.prl.clients"
+})
+public class CafcassSearchCaseApiConsumerApplication {
```
Comment (hmcts-kodatala) [inline src/contractTest/java/uk/gov/hmcts/reform/prl/clients/cafcass/CafcassSearchCaseApiConsumerApplication.java:10]: do we need this class , as it is empty. pls check?

File: src/contractTest/java/uk/gov/hmcts/reform/prl/clients/cafcass/CafcassApiConsumerTest.java [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+                                                                                    .stringType("PostCode", "HA11SA");
+                                                                        });
+                                                            });
+                                                })
+                                                .eachLike("respondents", (respondents) -> {
+                                                    respondents.stringType("id", "5581032c-1333-4bf2-9651-9d2433351c1a")
+                                                            .object("value", (respondentOne) -> {
+                                                                respondentOne.stringType("firstName", "sad")
+                                                                        .stringType("lastName", "sdf")
+                                                                        .stringType("gender", "male")
```
Comment (hmcts-kodatala) [inline src/contractTest/java/uk/gov/hmcts/reform/prl/clients/cafcass/CafcassApiConsumerTest.java:244]: are you doing pact test for ccd api search cases api or our api?

**Review by hmcts-kodatala — COMMENTED** [review]
Pls check review comments and we discussed to do pact test for integrations api like CCD API

File: src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/cafcass/CafCassControllerFunctionalTest.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        SearchResult expectedSearchResult = objectMapper.readValue(cafcassResponseStr, SearchResult.class);
+
+        Mockito.when(postcodeLookupService.isValidNationalPostCode(anyString(), anyString())).thenReturn(true);
+        Mockito.when(coreCaseDataApi.searchCases(anyString(), anyString(), anyString(), anyString())).thenReturn(expectedSearchResult);
+
+        MvcResult mvcResult = mockMvc.perform(get(SEARCH_CASE_ENDPOINT)
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .header(AUTHORISATION_HEADER, TEST_AUTH_TOKEN)
+                        .header(SERVICE_AUTHORISATION_HEADER, TEST_SERVICE_AUTH_TOKEN)
+                        .queryParam("start_date", "2022-08-22T10:44:43.49")
```
Comment (parvezhmcts) [inline src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/cafcass/CafCassControllerFunctionalTest.java:76]: use constants.

File: src/integrationTest/java/uk/gov/hmcts/reform/prl/controllers/cafcass/CafCassControllerIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+        Mockito.when(postcodeLookupService.isValidNationalPostCode(anyString(), eq(ENGLAND_POSTCODE_NATIONALCODE))).thenReturn(true);
+        Mockito.when(coreCaseDataApi.searchCases(anyString(), anyString(), anyString(), anyString())).thenReturn(searchResult);
+
+        mockMvc.perform(
+                        get(SEARCH_CASE_ENDPOINT)
+                                .contentType(APPLICATION_JSON)
+                                .header(AUTHORISATION_HEADER, TEST_AUTH_TOKEN)
+                                .header(SERVICE_AUTHORISATION_HEADER, TEST_SERVICE_AUTH_TOKEN)
+                                .queryParam("start_date", "2022-08-22T10:39:43.49")
```
Comment (parvezhmcts) [inline src/integrationTest/java/uk/gov/hmcts/reform/prl/controllers/cafcass/CafCassControllerIntegrationTest.java:74]: move to constants.

Comment (adetolacog) [reply src/integrationTest/java/uk/gov/hmcts/reform/prl/controllers/cafcass/CafCassControllerIntegrationTest.java:74]: Resolved

**Review by parvezhmcts — APPROVED** [review]
ok

**Review by hmcts-kodatala — APPROVED** [review]
OK
